ouroboros-network/src/Ouroboros/Network/Connections/Socket/Server.hs

```diff
@@ -0,0 +1,158 @@
+  -> IO t
+server bindaddr mkRequest k = bracket openSocket closeSocket $ \sock ->
+    k (acceptOne bindaddr sock mkRequest)
+  where
+
+  -- Use bracketOnError to ensure the socket is closed if any of the preparation
+  -- (setting options, bind, listen) fails. If not, the socket is returned
+  -- and the caller is responsible for closing it (the bracket at the top level
+  -- of this server definition).
+  openSocket :: IO Socket
+  openSocket = bracketOnError createSocket closeSocket $ \sock -> do
+    when isInet $ do
+      Socket.setSocketOption sock Socket.ReuseAddr 1
+      -- SO_REUSEPORT is not available on Windows.
+      -- But fortunately, SO_REUSEADDR on Windows does what SO_REUSEPORT does
+      -- on BSD-like implementations.
+#if !defined(mingw32_HOST_OS)
+      Socket.setSocketOption sock Socket.ReusePort 1
+#endif
+    when isInet6 $ Socket.setSocketOption sock Socket.IPv6Only 1
+    Socket.bind sock (forgetSockType bindaddr)
+    Socket.listen sock 1
+    return sock
+
+  createSocket :: IO Socket
+  createSocket = Socket.socket family Socket.Stream Socket.defaultProtocol
+
+  closeSocket :: Socket -> IO ()
+  closeSocket = Socket.close
+
+  isInet, isInet6 :: Bool
+  family :: Socket.Family
+  (isInet, isInet6, family) = case bindaddr of
+    SockAddrIPv4 _ _     -> (True,  False, Socket.AF_INET)
+    SockAddrIPv6 _ _ _ _ -> (True,  True,  Socket.AF_INET6)
+    SockAddrUnix _       -> (False, False, Socket.AF_UNIX)
+
+-- | Accepts one connection and include it, according to the parameter given
+-- (see the `Server` type synonym).
+-- Any exceptions thrown by accept will be re-thrown here, so be sure to
+-- handle them.
+acceptOne
+  :: SockAddr addrType -- Bind address; needed to construct ConnectionId
+  -> Socket
+  -> (SockAddr addrType -> request Remote)
+  -- ^ From the incoming socket address you must construct a remote request type
+  -> Server ConnectionId Socket IO request
+acceptOne bindaddr socket mkRequest = \includeConnection -> mask $ \restore -> do
+  (sock, addr) <- restore (Socket.accept socket)
+  case matchSockType bindaddr addr of
+    -- Should never happen.
+    Nothing -> error "mismatched socket address types"
+    -- In this case, matchSockType has shown us that bindaddr and peeraddr
+    -- have the same sockType parameter, so we can make a connection ID for
+    -- the pair.
+    Just peeraddr -> do
+      -- Note the order of the addresses in the identifier: we always put our
+      -- bind address first. The client (sibling module Client.hs) also puts
+      -- the bind address first.
+      let connId = makeConnectionId bindaddr peeraddr
+      -- Including the connection could fail exceptionally, in which case we
+      -- are still responsible for closing the socket.
+      includeResult <- restore (includeConnection connId sock (Socket.close sock) (mkRequest peeraddr))
+                       `onException`
+                       Socket.close sock
```

Comment:
Are you sure we are not closing a socket that has already been closed by `includeConnection`? It's not clear to me.

Reply:
it doesn't matter since `network` package protects us from that.